airflow/serialization/serialized_objects.py

```diff
@@ -653,10 +677,14 @@ def deserialize_operator(cls, encoded_op: Dict[str, Any]) -> BaseOperator:
                 v = cls._deserialize(v)
             # else use v as it is
 
-            setattr(op, k, v)
+            if hasattr(op, k) and isinstance(v, set):
+                getattr(op, k).update(v)
```

Comment:
What does this cover? Too hacky IMO.

Reply:
```python
    @property
    def upstream_task_ids(self) -> Set[str]:
        """@property: set of ids of tasks directly upstream"""
        return self._upstream_task_ids

    @property
    def downstream_task_ids(self) -> Set[str]:
        """@property: set of ids of tasks directly downstream"""
        return self._downstream_task_ids
```

Those can't be written but to the value of the set is mutable. So it's related to the `downstream_task_ids` vs `_downstream_task_ids` -- I think the change might be to make `downstream_task_id` a normal set attribute on BaseOperator (rather than have `_downstream_task_ids` on MappedOperator) -- I don't like storing/accessing the "private" fields here.

I'll see if that can be done without breaking the old serialization format.

Reply:
I'f changed this slightly to:

```
try:
  setattr(op, k, k)
except AttributeError:
   # Read only value, update in place
  getattr(op, k).update(v)
```

still a bit hacky, but marginally less so. I'll revisit this in the future.

Reply:
Does this affect anything other than down/upstream_task_ids? If not, maybe we should just rename the attribute on BaseOperator to remove the `_` (removing the public property, of course), and add a special case on deserialisation to migrate those attributes instead.

Reply:
Changed this (at tested that we still correctly inflate the old name correctly)

Reply:
Does the change in `deserialize_operator` also need to be updated?

Reply:
Which change? It's now this

```
            if k == "_downstream_task_ids":
                # Upgrade from old format/name
                k = "downstream_task_ids"
```

And then the set-except-get-update has done.

Reply:
Around here: https://github.com/apache/airflow/pull/20743/files#diff-807ca0a4fd53aeb41166621c9842b0f89b7931fc64e9a60befa36c776db45efaL661-L662

Reply:
Ah no, can't remove that one yet. I've added a TODO comment to refactor and remove it later.